### PR TITLE
[#1358] fix(spark): pre-check bytebuffer whether is direct before uncompress

### PR DESCRIPTION
**Pull request by zuston**
### What changes were proposed in this pull request?

precheck bytebuffer whether is direct before uncompress, and rebuild the temp bytebuffer.

### Why are the changes needed?

For #1358 . But this PR is not the fundamental solution

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing tests.

### CODE REVIEW

**Comment by leixm** [comment]
Can you explain why the original code fails to make the ByteBuffer consistent?

**Comment by zuston** [comment]
> Can you explain why the original code fails to make the ByteBuffer consistent?

I haven't found the root cause, but I think the bug is related with the `nioBuffer()` .

Anyway,  this PR is to use workaround to fix this. But we still need to dig this more.

**Comment by qqqttt123** [comment]
cc @LuciferYang Could you give us some advice about this issue?

File: client-spark/common/src/main/java/org/apache/spark/shuffle/reader/RssShuffleDataIterator.java [hunk 1]
```diff
@@ -150,14 +150,28 @@ public boolean hasNext() {
...
     long rawDataLength = rawData.limit() - rawData.position();
     totalRawBytesLength += rawDataLength;
     shuffleReadMetrics.incRemoteBytesRead(rawDataLength);
 
     int uncompressedLen = rawBlock.getUncompressLength();
     if (codec != null) {
-      if (uncompressedData == null || uncompressedData.capacity() < uncompressedLen) {
+      if (uncompressedData == null
+          || uncompressedData.capacity() < uncompressedLen
+          || !isSameMemoryType(uncompressedData, rawData)) {
```
Comment (jerqi) [inline client-spark/common/src/main/java/org/apache/spark/shuffle/reader/RssShuffleDataIterator.java:174]: Could you add some logs to get more information to improve the code later?

Comment (zuston) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/reader/RssShuffleDataIterator.java:174]: Yes, let me add it later.

File: client-spark/common/src/main/java/org/apache/spark/shuffle/reader/RssShuffleDataIterator.java [hunk 2]
```diff
@@ -150,14 +150,28 @@ public boolean hasNext() {
     return recordsIterator.hasNext();
   }
 
+  private boolean isSameMemoryType(ByteBuffer left, ByteBuffer right) {
```
Comment (LuciferYang) [inline client-spark/common/src/main/java/org/apache/spark/shuffle/reader/RssShuffleDataIterator.java:153]: how about `return left.isDirect() == right.isDirect()`?

Comment (zuston) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/reader/RssShuffleDataIterator.java:153]: !left.isDirect() && !right.isDirect() is also OK

File: client-spark/common/src/main/java/org/apache/spark/shuffle/reader/RssShuffleDataIterator.java (hunk 1 above)
Comment (zuston) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/reader/RssShuffleDataIterator.java:174]: Done

File: client-spark/common/src/main/java/org/apache/spark/shuffle/reader/RssShuffleDataIterator.java (hunk 2 above)
Comment (jerqi) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/reader/RssShuffleDataIterator.java:153]: `return left.isDirect() == right.isDirect()` includes the situation `!left.isDirect() && !right.isDirect()`.

Comment (zuston) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/reader/RssShuffleDataIterator.java:153]: Oh my fault!

Comment (LuciferYang) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/reader/RssShuffleDataIterator.java:153]: haha

File: client-spark/common/src/main/java/org/apache/spark/shuffle/reader/RssShuffleDataIterator.java [hunk 3]
```diff
@@ -150,14 +150,39 @@ public boolean hasNext() {
...
     int uncompressedLen = rawBlock.getUncompressLength();
     if (codec != null) {
-      if (uncompressedData == null || uncompressedData.capacity() < uncompressedLen) {
+      if (uncompressedData == null
+          || uncompressedData.capacity() < uncompressedLen
+          || !isSameMemoryType(uncompressedData, rawData)) {
+
+        if (LOG.isDebugEnabled()) {
+          if (!isSameMemoryType(uncompressedData, rawData)) {
+            LOG.debug(
```
Comment (jerqi) [inline client-spark/common/src/main/java/org/apache/spark/shuffle/reader/RssShuffleDataIterator.java:178]: I may suspect that readMemory and readLocalData will return different type bytebuffers.

Comment (zuston) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/reader/RssShuffleDataIterator.java:178]: I think No and have checked this happened on the https://github.com/apache/incubator-uniffle/issues/1358#issuecomment-1846626122

**Comment by zuston** [comment]
PTAL again @jerqi @LuciferYang

**Review by LuciferYang — APPROVED** [review]

**Comment by LuciferYang** [comment]
+1 from myside

**Comment by zuston** [comment]
Thanks for your review @jerqi @LuciferYang , Merged.

**Pull request merged**